### ISSUE TITLE
Amazon has removed "Order History  Reports" - but might be back again

### ISSUE THREAD

**Issue by LC1207** [jprouty/mint-amazon-tagger#67]
I've updated to the latest version of the tool and am seeing this error when I run the Tagger:

```
Logging into Amazon.com
Internal error while creating updates: Message: no such element: Unable to locate element: {"method":"css selector","selector":"[id="ap_password"]"}
  (Session info: chrome=84.0.4147.135)
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/mintamazontagger/main.py", line 607, in create_updates
    self.do_create_updates(args, parent)
  File "/usr/local/lib/python3.7/site-packages/mintamazontagger/main.py", line 651, in do_create_updates
    counter_progress_factory=progress_factory)
  File "/usr/local/lib/python3.7/site-packages/mintamazontagger/tagger.py", line 64, in create_updates
    progress_factory=indeterminate_progress_factory)
  File "/usr/local/lib/python3.7/site-packages/mintamazontagger/orderhistory.py", line 63, in fetch_order_history
    session_path=session_path)
  File "/usr/local/lib/python3.7/site-packages/mintamazontagger/orderhistory.py", line 175, in get_amzn_driver
    driver.find_element_by_id('ap_password').send_keys(
  File "/usr/local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 360, in find_element_by_id
    return self.find_element(by=By.ID, value=id_)
  File "/usr/local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 978, in find_element
    'value': value})['value']
  File "/usr/local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"css selector","selector":"[id="ap_password"]"}
  (Session info: chrome=84.0.4147.135)

Internal error while creating updates: Message: no such element: Unable to locate element: {"method":"css selector","selector":"[id="ap_password"]"}
  (Session info: chrome=84.0.4147.135)
```

I last ran this tool successfully about a month ago. Any help would be appreciated.

**Comment by justsubhash** [comment]
Looks like amazon removed the order reports link, possibly temporarily. It doesn't show for me anymore and I was trying to manually download them. See comments on this random site: https://www.idownloadblog.com/2020/02/06/amazon-export-order-history/

**Comment by jprouty** [comment]
Oh wow! Thanks for the heads up! This change makes the tagger dead in the water. There is a business API that's available, but I've tried to upgrade my personal account to try these out, but it doesn't seem easy going, and the API rate limits for historical orders are tragically low.

**Comment by briansea** [comment]
Following...

And also, seems support told a guy here it would be coming back? Hopefully!

https://www.reddit.com/r/amazonprime/comments/igmqvg/order_history_report_link_not_working/

**Comment by jbrady21** [comment]
@jprouty under the assumption that they don't revert the report access to the personal accounts, could a downloaded report from this be used instead?

https://chrome.google.com/webstore/detail/amazon-order-history-repo/mgkilgclilajckgnedgjgnfdokkgnibi/related?hl=en

**Comment by jprouty** [comment]
Hi @jbrady21, it certainly could, but would require a large re-work as well as a large loss in functionality. Specifically, that tool does not report:

- shipping/tracking information (which helps break up items per payments to ensure better matching)
- category information (for tagging the right categories/etc)
- per-item quantity and pricing breakdown (subtotal, tax, promo) (prevents order itemization in mint)

I'm currently taking a wait and see approach but may invest more time in the coming weeks.

**Comment by danlitwin** [comment]
![image](https://user-images.githubusercontent.com/1043716/92310919-8a197d00-ef80-11ea-9b16-be4b150e2a37.png)
Gone forever.  And I submitted a data request at that new page three days ago and still haven't received anything.

**Comment by jbrady21** [comment]
We’ve received and are processing your request to access your personal data.
We will provide your information to you as soon as we can. Usually, this should not take more than a month. In exceptional cases, for example if a request is more complex or if we are processing a high volume of requests, it might take longer, but if so we will notify you that there will be a delay.

One month?!?

Thanks,

James
Sent from my iPhone

On Sep 5, 2020, at 1:06 PM, Dan Litwin <notifications@github.com> wrote:

﻿

[image]<https://user-images.githubusercontent.com/1043716/92310919-8a197d00-ef80-11ea-9b16-be4b150e2a37.png>
Gone forever. And I submitted a data request at that new page three days ago and still haven't received anything.

**Comment by jprouty** [comment]
Thank you both for the legwork into figuring out what happened/replacements.

I am equally frustrated, so you're not alone! TBH: I'm not sure what to do just yet. Let's keep posting updates on tools/amazon's response as new information comes available.

**Comment by LC1207** [comment]
Just fyi - I requested my data from amazon and the turnaround for the csv's was about one day. This of course changes the workflow and perhaps also the logic required to process these. I was able to get all the files shown in screenshot. 

https://www.evernote.com/l/AU330jU3nq9MoaHa-Fj3W8FaEEKmAN6YU5c

**Comment by jprouty** [comment]
@LC1207 Do you mind sharing how you got that data? What exactly did you ask for? And what channel (email/chat/phone)? Was it a GDPR ask?

Also, if you could share the first row of each csv file, that'd be helpful for me to understand how we might map the new format and determine if it's viable. If you're okay, even including a single row of data would be useful too (feel free to scramble/remove all string/addresses/payments/etc). Thanks!

**Comment by LC1207** [comment]
No problem - go here https://www.amazon.com/gp/privacycentral/dsar/preview.html?ie=UTF8&%2AVersion%2A=1&%2Aentries%2A=0& and then select "Your Orders" and you can put in your request. I imagine this is driven by GDPR / CCPA. Once you get all your data you should see it all.

I can try to get you some rows tomorrow too.

**Comment by LC1207** [comment]
Here's an example - https://docs.google.com/spreadsheets/d/1vtCPcT-Q8T5kxM04UW8d1tQdAJRsELpyNLPexJyipGk/edit?usp=sharing

**Comment by jprouty** [comment]
Neat, I requested my data and received results 7 days later :-P Thanks @LC1207 for the sample; I'll inspect my dataset as well and see if there's any immediate next steps with this new format.

**Comment by felciano** [comment]
Possibly helpful in debugging: Chrome extension that tries to scrape the data from the UI itself:

https://chrome.google.com/webstore/detail/amazon-order-history-repo/mgkilgclilajckgnedgjgnfdokkgnibi?hl=en

**Comment by jprouty** [comment]
Absolutely; scraping individual order pages is probably the best option we have. I tried the extension and was somewhat surprised how quickly it went without getting rate limited.

At one point I was looking into [this API](http://docs.developer.amazonservices.com/en_US/orders-2013-09-01/Orders_ListOrders.html) and [the matching GetOrder](http://docs.developer.amazonservices.com/en_US/orders-2013-09-01/Orders_GetOrder.html), but the default quota/restore rate is far to low to be useful (but maybe not if coupled with a local database). 

I possibly have some free time in the near future (next week or so) and will take a closer look then. Stay tuned!

**Comment by felciano** [comment]
One thing that might be worth considering is some light decoupling in the flow. The ability to upload split transaction details to Mint in a scalable robust fashion has standalone value. So it might be worth decoupling that through a well-defined transaction file format that could be handcrafted if necessary. That way if things break again on the Amazon side, users can still opt to create a detailed Amazon transactions file by hand, and then have your app upload the whole batch automatically. 

Obviously, it is nicer if the round trip can be automated, but that way there is a fallback such that Amazon redesigns won't break the whole flow.

**Comment by turnepf** [comment]
Wow, this is like finding out about a great new restaurant in town only to visit and discover it closed the night before. Thanks for the work on this, even if it appears to be dead in the water. Would've been awesome.

**Comment by kwylie** [comment]
@jprouty ... anything we can do to help you get this up and running again?  It was such a time-saver!

**Comment by jprouty** [comment]
@turnepf Thanks for the kind words and sorry for the bad timing!

@kwylie Great question. It's been two months since Amazon removed order reports, so I think it's safe to say it's not coming back. I am abandoning this tool as it stands today, as a desktop python app and works only between Amazon/Mint.

I am actively working to derisk the Chrome extension path. Specifically, I'm considering a new "My purchases" Chrome extension, which:
- Maintains a secure database of itemized purchases from Amazon
- Scrapes purchase details from relevant merchant on demand OR **[Optional]** a regular basis. Scraping would work similar to [this extension](https://chrome.google.com/webstore/detail/amazon-order-history-repo/mgkilgclilajckgnedgjgnfdokkgnibi/related?hl=en)
- **[New/Optional]** Adds new overlay feature to financial institutions and tools. e.g. log into your bank website, look at your transaction history page, and if it finds a matching purchase in your database, creates an overlay/tooltip that shows purchase info.
- **[Optional]** Writes/persists matched purchases to institutions/tools (very much like this tagger tool).
- Extends to include other merchants (e.g. Target, BestBuy, DigiKey, etc)
- Extends to include other institutions/tools (e.g. your bank, Mint, Personal Capital)

This is a complete rewrite and will live in a new repository; I'll post updates here as things come into focus! It should be expected that this rewrite may take until end of year.

**Comment by jprouty** [comment]
@kwylie Final note; I'm support of anyone forking this project and attempting to get the existing logic working based on the output of [this extension](https://chrome.google.com/webstore/detail/amazon-order-history-repo/mgkilgclilajckgnedgjgnfdokkgnibi/related?hl=en). This may not be straight forward as the matching logic is pretty detailed and depends on a bunch of AMZN quicks in how they itemize reports, but it should be doable (but also a bit of a dead-end/temporary patchwork).

**Comment by jbrady21** [comment]
Jeff,

If your looking, put me down as a QA tester whenever needed. I should have time to perform any testing that comes up every evening.

Thanks,

James
Sent from my iPhone

On Oct 24, 2020, at 12:14 PM, Jeff Prouty <notifications@github.com> wrote:

﻿

@kwylie<https://github.com/kwylie> Final note; I'm support of anyone forking this project and attempting to get the existing logic working based on the output of this extension<https://chrome.google.com/webstore/detail/amazon-order-history-repo/mgkilgclilajckgnedgjgnfdokkgnibi/related?hl=en>. This may not be straight forward as the matching logic is pretty detailed and depends on a bunch of AMZN quicks in how they itemize reports, but it should be doable (but also a bit of a dead-end/temporary patchwork).

**Comment by ghostlybao** [comment]
does anyone have generic samples of the Items, Orders, and Refunds CSV file that the tool uses. This way I can at least take the data dump Amazon gives and conform it to the file formats

**Comment by jprouty** [comment]
Take a look at this: [ExampleAMZNOrderHistory.zip](https://github.com/jprouty/mint-amazon-tagger/files/5440031/ExampleAMZNOrderHistory.zip)

The unit tests should also be helpful.

**Comment by kwylie** [comment]
How's the rewrite going?

**Comment by carson-51** [comment]
FYI, I was able to download these manually from Amazon and then use the App to tag in Mint. I found the link here: https://www.amazon.com/gp/b2b/reports/?ie=UTF8&download-report.x=1&isInternal=0&js=1&reportId=A2JK9NBNFLARNF

**Comment by jprouty** [comment]
Whoa! Nice; that'll get things going again! That's the same URL as before. It appears AMZN is no longer linking to it from "your account" or "order history" pages. This should resolve the original issue at top and require no update to any of the auto fetching logic that's already in the wild. I'll verify a bit later.

@kwylie The rewrite is going slowly :) Had a new kid and expected this to be the in between project, but instead ended up doing a bunch of house projects. Assuming the page is back, I'll probably deprioritize the rewrite a bit.

**Comment by dev415** [comment]
Great news! The only issue I am seeing is that the late app throws an error because Chrome is no longer version 84.

![image](https://user-images.githubusercontent.com/53407279/102038029-3f37ae80-3d94-11eb-886f-d2f428104c02.png)

**Comment by danlitwin** [comment]
> The rewrite is going slowly :) Had a new kid 

Congrats! Sounds like you′ll really need to get the script going again for your own sake. When I had my first last year, this tagger was a godsend for tracking expenses in my Baby Supplies category.

**Comment by jprouty** [comment]
Thanks @dev415 and @danlitwin about the kid! So far so smooth 🤞 

@dev415 I'm seeing the same thing (different version). I thought there was a quick fix but it's looking a bit more involved than expected.

**Comment by jprouty** [comment]
Okay, I think I've resolved the chromedriver issue. Please take a go on the latest release, pypi, or from source. 1.31 should be operable :) We're back in business!

**Comment by dev415** [comment]
Thanks for the quick turnaround as always. Looks like we still might have an issue.

![image](https://user-images.githubusercontent.com/53407279/102291092-51d6f280-3f10-11eb-8555-3b93b1504dbe.png)

**Comment by kwylie** [comment]
Amazing!  Got it working again.  Thanks @jprouty !!!

**Comment by mwong56** [comment]
Can confirm its working for me too! I am getting an error on the logout which isn't too big of a deal 

```
All done; no new tags to be updated at this point in time!
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/michael/.local/lib/python3.7/site-packages/mintapi/api.py", line 476, in close
    self.driver.find_element_by_id('link-logout').click()
  File "/home/michael/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 80, in click
    self._execute(Command.CLICK_ELEMENT)
  File "/home/michael/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 633, in _execute
    return self._parent.execute(command, params)
  File "/home/michael/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/michael/.local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.ElementClickInterceptedException: Message: element click intercepted: Element <a id="link-logout" class="navbar-link" href="/logout.event?task=explicit&amp;country=">...</a> is not clickable at point (965, 31). Other element would receive the click: <div data-automation-id="ModalDialog--container" open="" role="dialog" class="StyledModalDialog__Container-cn89yw-1 iIlfZT" style="z-index: 102;">...</div>
  (Session info: chrome=87.0.4280.88)

```

**Comment by jprouty** [comment]
@dev415 I'm able to replicate the behavior on win10. I'll take a look and update this thread. The culprit code is in the mintapi library, so a fix may require a bit more time/review to get it in.

**Comment by jprouty** [comment]
@dev415 I think I have a workaround. Do you mind trying again? I assume you're using a [released binary from 1.31 page](https://github.com/jprouty/mint-amazon-tagger/releases/tag/1.31) as I'm unable to replicate from source (only once packaged up). I've updated the win10 binary for 1.31, so please try redownloading and see if that resolves the issue!

**Comment by dev415** [comment]
@jprouty It worked! We are back in business!

**Comment by jprouty** [comment]
That's great @dev415 !

@mwong56 your issue is that there's a modal (aka a pop up dialog box) that mint is showing you. It's intercepting all clicks until you dismiss the box. I got a modal like this a year back or so and decided that it makes the most sense for mintapi to do nothing about. ie, if the app closes the modal on your behalf, you'll never see that information it's trying to show. The side effect is not logging out cleanly; however, the session will go stale (just as if you forgot to logout from mint in your own browser before closing the window/tab).

**Comment by joelvaneenwyk** [comment]
@jprouty, FYI, tested this on my side today directly from `master` branch and it all worked swimmingly. Thank you!

**Comment by ijoseph** [comment]
Worked for me after quite a bit of trial and error. (But still not even in the same league as having to do that manually -- thanks again!). 

I'm on MacOS Catalina, and chromedriver was definitely acting out... eventually, I realized I had to `rm ~/chromedriver` every time before starting it, otherwise the Qt UI would hang apparently waiting for chromedriver, but the chromedriver never appeared. This was especially problematic because after the Amazon chromedriver session completed, it was not able to begin another one for the Mint session. Anyway, probably unrelated issues, but I thought I'd mention regardless.

**Comment by jprouty** [comment]
Hi @ijoseph. Apologies for the difficulty! I'm going to try and replicate the issue on my laptop. Hopefully it will happen on Big Sur as well.

Were you using:

1. The pre-built binary from github releases page
2. Via pypi e.g. `pip install`
3. From source e.g. `git clone`

If you're using (3), you might try upgrading all dependencies `pip install --upgrade -r requirements`.
(2) should complain at install/upgrade time (try pip install --upgrade mint-amazon-tagger`)
If you haven't tried (1), and aren't opposed to doing so, I'd be interested in knowing how that workings for you.

Thank you!

**Comment by ijoseph** [comment]
Thanks @jprouty ! 

Sorry, forgot to mention I was using `master` (from `git clone`). 

Ah right, that may have been the issue (not upgrading requirements) -- I don't believe I did that.

**Comment by jprouty** [comment]
Thanks @ijoseph . I had some messed up logic in the chromedriver selection code that I did not test on mac. It's all good now as of 1.32 (or master from source).

I am going to close this issue for now, as things seem to be stable. Please create a new issue if anyone runs into any further difficulties or issues getting the tool up and running.
